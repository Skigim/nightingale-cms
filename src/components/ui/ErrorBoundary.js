/**
 * ErrorBoundary.js - Error boundary UI component
 *
 * Class component for graceful error handling in React applications.
 * Catches JavaScript errors anywhere in the child component tree and displays a fallback UI.
 *
 * Note: This is a class component as Error Boundaries require class components in React.
 *
 * @namespace NightingaleUI
 * @version 1.0.0
 * @author Nightingale CMS Team
 */

/**
 * ErrorBoundary Component
 * Wraps components to catch and handle errors gracefully
 *
 * @class ErrorBoundary
 * @extends React.Component
 */
class ErrorBoundary extends window.React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false, error: null, errorInfo: null };
  }

  // eslint-disable-next-line no-unused-vars
  static getDerivedStateFromError(error) {
    // Update state so the next render will show the fallback UI
    return { hasError: true };
  }

  componentDidCatch(error, errorInfo) {
    // Log error details
    this.setState({
      error: error,
      errorInfo: errorInfo,
    });
    // (logging removed) hook for future logger integration
  }

  render() {
    // React safety check (though this should always be available in class components)
    if (!window.React) {
      return null;
    }

    const e = window.React.createElement;

    if (this.state.hasError) {
      // Fallback UI when an error occurs
      return e(
        'div',
        {
          className:
            'h-full w-full flex items-center justify-center bg-gray-900 text-white p-8',
        },
        e(
          'div',
          { className: 'max-w-2xl text-center' },
          e(
            'h1',
            { className: 'text-3xl font-bold mb-4' },
            'Something went wrong',
          ),
          e(
            'p',
            { className: 'text-gray-300 mb-6' },
            "The application encountered an error and couldn't recover.",
          ),
          e(
            'button',
            {
              className:
                'bg-blue-600 hover:bg-blue-700 px-6 py-2 rounded text-white font-medium',
              onClick: () => window.location.reload(),
            },
            'Reload Application',
          ),
          this.state.error &&
            e(
              'details',
              { className: 'mt-6 text-left' },
              e(
                'summary',
                { className: 'cursor-pointer text-gray-400' },
                'Technical Details',
              ),
              e(
                'pre',
                {
                  className:
                    'mt-2 text-sm bg-gray-800 p-4 rounded overflow-auto',
                },
                this.state.error.toString() +
                  '\n' +
                  (this.state.errorInfo?.componentStack || ''),
              ),
            ),
        ),
      );
    }

    // When there's no error, render children normally
    return this.props.children;
  }
}

// PropTypes for validation
if (typeof window !== 'undefined' && window.PropTypes) {
  ErrorBoundary.propTypes = {
    children: window.PropTypes.node.isRequired,
  };
}

// Self-registration for both module and script loading
if (typeof window !== 'undefined') {
  // Register globally for backward compatibility
  window.ErrorBoundary = ErrorBoundary;

  // Register with NightingaleUI registry if available
  if (window.NightingaleUI) {
    window.NightingaleUI.registerComponent(
      'ErrorBoundary',
      ErrorBoundary,
      'core',
    );
  }
}

// Export for ES6 module compatibility
if (typeof module !== 'undefined' && module.exports) {
  module.exports = ErrorBoundary;
}

// ES6 Module Export
export default ErrorBoundary;
