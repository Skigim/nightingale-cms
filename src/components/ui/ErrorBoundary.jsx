import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { registerComponent } from '../../services/registry';

/**
 * ErrorBoundary.jsx - Error boundary UI component
 *
 * Class component for graceful error handling in React applications.
 * Catches JavaScript errors anywhere in the child component tree and displays a fallback UI.
 *
 * Note: This is a class component as Error Boundaries require class components in React.
 *
 * @version 1.0.0
 * @author Nightingale CMS Team
 */

/**
 * ErrorBoundary Component
 * Wraps components to catch and handle errors gracefully
 *
 * @class ErrorBoundary
 * @extends React.Component
 */
class ErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false, error: null, errorInfo: null };
  }

  // eslint-disable-next-line no-unused-vars
  static getDerivedStateFromError(error) {
    // Update state so the next render will show the fallback UI
    return { hasError: true };
  }

  componentDidCatch(error, errorInfo) {
    // Log error details
    this.setState({
      error: error,
      errorInfo: errorInfo,
    });

    // Log via Nightingale logger if available
    try {
      const logger = window.NightingaleLogger?.get('ErrorBoundary');
      logger?.error('Error caught by boundary:', error);
    } catch (logError) {
      // Silently handle logger errors
      console.error('Failed to log error:', logError);
    }
  }

  componentDidUpdate(prevProps) {
    // Reset error state when key prop changes
    // React does not pass 'key' through props; to allow manual resets, use a custom prop like 'boundaryKey'
    if (
      this.props.boundaryKey !== prevProps.boundaryKey &&
      this.state.hasError
    ) {
      this.setState({ hasError: false, error: null, errorInfo: null });
    }
  }

  render() {
    if (this.state.hasError) {
      // Fallback UI when an error occurs
      return (
        <div className="h-full w-full flex items-center justify-center bg-gray-900 text-white p-8">
          <div className="max-w-2xl text-center">
            <h1 className="text-3xl font-bold mb-4">Something went wrong</h1>
            <p className="text-gray-300 mb-6">
              The application encountered an error and couldn&apos;t recover.
            </p>
            <button
              className="bg-blue-600 hover:bg-blue-700 px-6 py-2 rounded text-white font-medium"
              onClick={() => window.location.reload()}
            >
              Reload Application
            </button>
            {this.state.error && (
              <details className="mt-6 text-left">
                <summary className="cursor-pointer text-gray-400">
                  Technical Details
                </summary>
                <pre className="mt-2 text-sm bg-gray-800 p-4 rounded overflow-auto">
                  {this.state.error.toString() +
                    '\n' +
                    (this.state.errorInfo?.componentStack || '')}
                </pre>
              </details>
            )}
          </div>
        </div>
      );
    }

    // When there's no error, render children normally
    return this.props.children;
  }
}

// PropTypes for validation
ErrorBoundary.propTypes = {
  children: PropTypes.node.isRequired,
  boundaryKey: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

// Self-registration for both module and script loading
// Register with UI registry (legacy global removal)
registerComponent('ui', 'ErrorBoundary', ErrorBoundary);

// Export for ES6 module compatibility
if (typeof module !== 'undefined' && module.exports) {
  module.exports = ErrorBoundary;
}

// ES6 Module Export
export default ErrorBoundary;
