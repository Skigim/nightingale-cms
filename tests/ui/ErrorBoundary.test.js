/**
 * @jest-environment jsdom
 */

import { render, screen, fireEvent } from '@testing-library/react';
import React from 'react';
import ErrorBoundary from '../../src/components/ui/ErrorBoundary.jsx';

// Mock component that throws an error when instructed
const ThrowError = ({ shouldThrow }) => {
  if (shouldThrow) {
    throw new Error('Test error from ThrowError component');
  }
  return <div>No error</div>;
};

// Mock component that always works
const WorkingComponent = () => <div>Working component</div>;

describe('ErrorBoundary Component', () => {
  // Suppress console.error for these tests since we're intentionally causing errors
  const originalError = console.error;
  beforeAll(() => {
    console.error = jest.fn();
  });

  afterAll(() => {
    console.error = originalError;
  });

  beforeEach(() => {
    // Mock window.location.reload
    Object.defineProperty(window, 'location', {
      value: {
        reload: jest.fn(),
      },
      writable: true,
    });
  });

  test('renders children when no error occurs', () => {
    render(
      <ErrorBoundary>
        <WorkingComponent />
      </ErrorBoundary>,
    );

    expect(screen.getByText('Working component')).toBeInTheDocument();
  });

  test('renders error UI when child component throws error', () => {
    render(
      <ErrorBoundary>
        <ThrowError shouldThrow={true} />
      </ErrorBoundary>,
    );

    expect(screen.getByText('Something went wrong')).toBeInTheDocument();
    expect(
      screen.getByText(/The application encountered an error/),
    ).toBeInTheDocument();
    expect(screen.getByText('Reload Application')).toBeInTheDocument();
  });

  test('shows technical details when expanded', () => {
    render(
      <ErrorBoundary>
        <ThrowError shouldThrow={true} />
      </ErrorBoundary>,
    );

    const details = screen.getByText('Technical Details');
    expect(details).toBeInTheDocument();

    // Click to expand details
    fireEvent.click(details);

    // Should show error details in a pre element
    const errorDetails = screen.getByText(
      /Test error from ThrowError component/,
    );
    expect(errorDetails).toBeInTheDocument();
    expect(errorDetails.tagName).toBe('PRE');
  });

  test('reload button calls window.location.reload', () => {
    render(
      <ErrorBoundary>
        <ThrowError shouldThrow={true} />
      </ErrorBoundary>,
    );

    const reloadButton = screen.getByText('Reload Application');
    fireEvent.click(reloadButton);

    expect(window.location.reload).toHaveBeenCalledTimes(1);
  });

  test('has correct styling for error UI', () => {
    const { container } = render(
      <ErrorBoundary>
        <ThrowError shouldThrow={true} />
      </ErrorBoundary>,
    );

    const errorContainer = container.querySelector('.bg-gray-900');
    expect(errorContainer).toBeInTheDocument();
    expect(errorContainer).toHaveClass(
      'h-full',
      'w-full',
      'flex',
      'items-center',
      'justify-center',
      'bg-gray-900',
      'text-white',
      'p-8',
    );
  });

  test('error boundary catches errors during rendering', () => {
    const { rerender } = render(
      <ErrorBoundary>
        <ThrowError shouldThrow={false} />
      </ErrorBoundary>,
    );

    // Initially no error
    expect(screen.getByText('No error')).toBeInTheDocument();

    // Trigger error by re-rendering with shouldThrow=true
    rerender(
      <ErrorBoundary>
        <ThrowError shouldThrow={true} />
      </ErrorBoundary>,
    );

    // Should now show error boundary UI
    expect(screen.getByText('Something went wrong')).toBeInTheDocument();
    expect(screen.queryByText('No error')).not.toBeInTheDocument();
  });

  test('error boundary handles componentDidCatch lifecycle', () => {
    const errorBoundary = render(
      <ErrorBoundary>
        <ThrowError shouldThrow={true} />
      </ErrorBoundary>,
    );

    // Should render error UI, indicating componentDidCatch was called
    expect(screen.getByText('Something went wrong')).toBeInTheDocument();
  });

  test('renders with proper accessibility structure', () => {
    render(
      <ErrorBoundary>
        <ThrowError shouldThrow={true} />
      </ErrorBoundary>,
    );

    // Check for proper heading structure
    const heading = screen.getByRole('heading', { level: 1 });
    expect(heading).toHaveTextContent('Something went wrong');

    // Check for button
    const button = screen.getByRole('button', { name: 'Reload Application' });
    expect(button).toBeInTheDocument();

    // Check for details/summary structure
    const details = screen.getByText('Technical Details');
    expect(details.tagName).toBe('SUMMARY');
  });
});
