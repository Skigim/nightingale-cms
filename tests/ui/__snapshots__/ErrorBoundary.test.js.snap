// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary Component matches snapshot for error state 1`] = `
<div
  class="h-full w-full flex items-center justify-center bg-gray-900 text-white p-8"
>
  <div
    class="max-w-2xl text-center"
  >
    <h1
      class="text-3xl font-bold mb-4"
    >
      Something went wrong
    </h1>
    <p
      class="text-gray-300 mb-6"
    >
      The application encountered an error and couldn't recover.
    </p>
    <button
      class="bg-blue-600 hover:bg-blue-700 px-6 py-2 rounded text-white font-medium"
    >
      Reload Application
    </button>
    <details
      class="mt-6 text-left"
    >
      <summary
        class="cursor-pointer text-gray-400"
      >
        Technical Details
      </summary>
      <pre
        class="mt-2 text-sm bg-gray-800 p-4 rounded overflow-auto"
      >
        Error: Snapshot test error

    at ThrowError (/home/runner/work/nightingale-cms/nightingale-cms/src/components/ui/__tests__/ErrorBoundary.test.js:166:13)
    at ErrorBoundary (/home/runner/work/nightingale-cms/nightingale-cms/src/components/ui/ErrorBoundary.js:23:5)
      </pre>
    </details>
  </div>
</div>
`;

exports[`ErrorBoundary Component matches snapshot for normal state 1`] = `
<div
  class="test-child"
>
  Normal content
</div>
`;
